### PR TITLE
OIDC

### PR DESCRIPTION
**Pull request by minottic**
## Description

First version of the OIDC implementation. Current limits: 

1. it supports one single OIDC provider
2. it does not yet support custom callback logic
3. tests are missing

## Motivation
issue: #107 

## Tests included/Docs Updated?

- [ ] Included for each change/fix?
- [ ] Passing? (Merge will not be approved unless this is checked) 
- [ ] Docs updated?
- [ ] New packages used/requires npm install? 
- [ ] Toggle added for new features?

### CODE REVIEW

**Comment by minottic** [comment]
@dylanmcreynolds , @nitrosx I've pushed what I've done before I leave for holiday, even though it still has quite some limitations. I think this first version could in principle be merged after adding the tests and when happy. 

The improvements (1,2) IMHO can be added in later PRs

File: src/auth/strategies/oidc.strategy.ts [hunk 1]
```diff
@@ -0,0 +1,141 @@
...
+  Client,
+  UserinfoResponse,
+  TokenSet,
+  Issuer,
+} from "openid-client";
+import { AuthService } from "../auth.service";
+import { Profile } from "passport";
+import { UserProfile } from "src/users/schemas/user-profile.schema";
+
+export class buildOpenIdClient {
```
Comment (dylanmcreynolds) [inline src/auth/strategies/oidc.strategy.ts:20]: Super picky point, classes should be UpperCamelCase

File: src/auth/strategies/oidc.strategy.ts [hunk 2]
```diff
@@ -0,0 +1,141 @@
...
+  constructor(private configService: ConfigService) {}
+  async build() {
+    const oidcConfig = this.configService.get<Record<string, unknown>>("oidc");
+    const TrustIssuer = await Issuer.discover(
```
Comment (dylanmcreynolds) [inline src/auth/strategies/oidc.strategy.ts:24]: lowerCamelCase

File: src/auth/strategies/oidc.strategy.ts [hunk 3]
```diff
@@ -0,0 +1,141 @@
...
+    if (userinfo.preferred_username) {
+      profile.username = userinfo.preferred_username;
+    }
+    if (userinfo.email) {
+      profile.emails = [{ value: userinfo.email }];
+      profile.email = userinfo.email;
+    }
+
+    profile.accessGroups = userinfo[
+      this.configService.get("oidc").accessGroups
```
Comment (dylanmcreynolds) [inline src/auth/strategies/oidc.strategy.ts:136]: Does this mean that access groups are always hard-coded in configuration?

**Comment by dylanmcreynolds** [comment]
I may be missing somethbing, but It seems to me that to be useful, we also need some mechanism of asking an external system for a user's group memberships (as the old backend got with the login-callback.js). In your case, are you getting that right out of keycloak in the profile? In this PR, it looks to me that it's hard-coded from the configuration. So, again, if I'm reading this right, the PR solves authentication with OIDC but not authorization, no?

File: src/auth/strategies/oidc.strategy.ts (hunk 3 above)
Comment (minottic) [reply src/auth/strategies/oidc.strategy.ts:136]: only the key to retrieve the accessGroups values from the userInfo object (which is returned by the oidc when calling the userInfo endpoint):
Schematically, these are the steps: 

1. the code gets the oidc `.well-know` config
2. authentication is triggered through the oidc provider
3. when successful, oidc returns a token (jwt) 
4. token from 3 is used to call the oidc userInfo endpoint, which is contained in 1
5. 4 returns general user information, e.g. name, username etc and here this PR makes the assumption that one property of the userInfo is the accessGroups
6. line 136 extracts the accessGroups having specified the name of the userinfo accessGroups property in the config (e.g. at PSI we call them p-groups)

**Comment by minottic** [comment]
to some extent, yes, and that is what I tried to explain in comments (1, 2) opening the PR. 

For this PR to be useful, there is the assumption that the accessGroups are in the userInfo which are returned after the oidc authentication (see comment above).

I agree that this PR does not address the case where accessGroups come from different sources neither it covers cases where there are multiple oidc providers, but I think, after having resolved your other comments (which I'll do asap), that these 2 limitations can be tackled in later PRs

**Review by dylanmcreynolds — APPROVED** [review]
For some sites, a follow up PR will be required to obtain user group from an external source. But this will work for sites for whom that information is coming from the OIDC provider. Great stuff. Thanks!

**Comment by minottic** [comment]
This is still open as I was hoping to add some acceptance tests, but I have struggled to find time recently. My idea was to add to the test docker-compose an instance of keycloak with some predefined accounts that can be used to test

**Pull request merged**